### ISSUE TITLE
How to convert Total-text dataset for the training of DB?

### ISSUE THREAD

**Issue by tzm-tora** [open-mmlab/mmocr#1926]
### Branch

master branch  https://mmocr.readthedocs.io/en/latest/

### 📚 The doc issue

The ICDAR2015 dataset converter script (detext_converter.py) is available. In the training of Total-text, I first tried to make the annotation format of Total-text into the original IC15 format, then I used the detext_converter.py to convert them to the training JSON file. But my experiments got:
icdar/precision: 0.7730  icdar/recall: 0.7396  icdar/hmean: 0.7559  

This result is much lower than your report. I think the training JSON file of the labels maybe not convert correctly.
so, could you provide a total_converter.py?
or at least provide a mini_totaltext dataset to show us what is the correct format of the curved text dataset?

my testing log is as follows:

System environment:
    sys.platform: linux
    Python: 3.8.16 (default, Mar  2 2023, 03:21:46) [GCC 11.2.0]
    CUDA available: True
    numpy_random_seed: 1734280501
    GPU 0: NVIDIA Graphics Device
    CUDA_HOME: /usr/local/cuda
    NVCC: Cuda compilation tools, release 11.6, V11.6.124
    GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
    PyTorch: 1.10.0
    PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

    TorchVision: 0.11.0
    OpenCV: 4.7.0
    MMEngine: 0.7.4

Runtime environment:
    cudnn_benchmark: False
    mp_cfg: {'mp_start_method': 'fork', 'opencv_num_threads': 0}
    dist_cfg: {'backend': 'nccl'}
    seed: 1734280501
    Distributed launcher: none
    Distributed training: False
    GPU number: 1

2023/06/13 16:32:47 - mmengine - INFO - Config:
model = dict(
    type='DBNet',
    backbone=dict(
        type='mmdet.ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=-1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=False,
        style='pytorch',
        dcn=dict(type='DCNv2', deform_groups=1, fallback_on_stride=False),
        init_cfg=dict(type='Pretrained', checkpoint='torchvision://resnet50'),
        stage_with_dcn=(False, True, True, True)),
    neck=dict(
        type='FPNC', in_channels=[256, 512, 1024, 2048], lateral_channels=256),
    det_head=dict(
        type='DBHead',
        in_channels=256,
        module_loss=dict(type='DBModuleLoss'),
        postprocessor=dict(type='DBPostprocessor', text_repr_type='quad')),
    data_preprocessor=dict(
        type='TextDetDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True,
        pad_size_divisor=32))
train_pipeline = [
    dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
    dict(
        type='LoadOCRAnnotations',
        with_polygon=True,
        with_bbox=True,
        with_label=True),
    dict(type='FixInvalidPolygon', min_poly_points=4),
    dict(
        type='TorchVisionWrapper',
        op='ColorJitter',
        brightness=0.12549019607843137,
        saturation=0.5),
    dict(
        type='ImgAugWrapper',
        args=[['Fliplr', 0.5], {
            'cls': 'Affine',
            'rotate': [-10, 10]
        }, ['Resize', [0.5, 3.0]]]),
    dict(type='RandomCrop', min_side_ratio=0.1),
    dict(type='Resize', scale=(640, 640), keep_ratio=True),
    dict(type='Pad', size=(640, 640)),
    dict(
        type='PackTextDetInputs',
        meta_keys=('img_path', 'ori_shape', 'img_shape'))
]
test_pipeline = [
    dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
    dict(type='Resize', scale=(1333, 736), keep_ratio=True),
    dict(
        type='LoadOCRAnnotations',
        with_polygon=True,
        with_bbox=True,
        with_label=True),
    dict(type='FixInvalidPolygon', min_poly_points=4),
    dict(
        type='PackTextDetInputs',
        meta_keys=('img_path', 'ori_shape', 'img_shape', 'scale_factor'))
]
train_data_root = 'data/TT_train'
test_data_root = 'data/TT_test'
totaltext_textdet_train = dict(
    type='OCRDataset',
    data_root='data/TT_train',
    ann_file='instances_training.json',
    filter_cfg=dict(filter_empty_gt=True, min_size=32),
    pipeline=[
        dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
        dict(
            type='LoadOCRAnnotations',
            with_polygon=True,
            with_bbox=True,
            with_label=True),
        dict(type='FixInvalidPolygon', min_poly_points=4),
        dict(
            type='TorchVisionWrapper',
            op='ColorJitter',
            brightness=0.12549019607843137,
            saturation=0.5),
        dict(
            type='ImgAugWrapper',
            args=[['Fliplr', 0.5], {
                'cls': 'Affine',
                'rotate': [-10, 10]
            }, ['Resize', [0.5, 3.0]]]),
        dict(type='RandomCrop', min_side_ratio=0.1),
        dict(type='Resize', scale=(640, 640), keep_ratio=True),
        dict(type='Pad', size=(640, 640)),
        dict(
            type='PackTextDetInputs',
            meta_keys=('img_path', 'ori_shape', 'img_shape'))
    ])
totaltext_textdet_test = dict(
    type='OCRDataset',
    data_root='data/TT_test',
    ann_file='instances_test.json',
    test_mode=True,
    pipeline=[
        dict(type='LoadImageFromFile', color_type='color_ignore_orientation'),
        dict(type='Resize', scale=(1333, 736), keep_ratio=True),
        dict(
            type='LoadOCRAnnotations',
            with_polygon=True,
            with_bbox=True,
            with_label=True),
        dict(type='FixInvalidPolygon', min_poly_points=4),
        dict(
            type='PackTextDetInputs',
            meta_keys=('img_path', 'ori_shape', 'img_shape', 'scale_factor'))
    ])
default_scope = 'mmocr'
env_cfg = dict(
    cudnn_benchmark=False,
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    dist_cfg=dict(backend='nccl'))
randomness = dict(seed=None)
default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=100),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(type='CheckpointHook', interval=20),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    sync_buffer=dict(type='SyncBuffersHook'),
    visualization=dict(
        type='VisualizationHook',
        interval=1,
        enable=True,
        show=False,
        draw_gt=True,
        draw_pred=True))
log_level = 'INFO'
log_processor = dict(type='LogProcessor', window_size=10, by_epoch=True)
load_from = '_ckpt_new/TT/epoch_1200.pth'
resume = False
val_evaluator = dict(type='HmeanIOUMetric')
test_evaluator = [
    dict(type='HmeanIOUMetric'),
    dict(
        type='DumpResults',
        out_file_path='_test_test/TT/epoch_1200.pth_predictions.pkl')
]
vis_backends = [dict(type='LocalVisBackend')]
visualizer = dict(
    type='TextDetLocalVisualizer',
    name='visualizer',
    vis_backends=[dict(type='LocalVisBackend')],
    save_dir='_test_test/TT/show_dir')
epoch = 1200
val_interval = 20.0
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(type='SGD', lr=0.007, momentum=0.9, weight_decay=0.0001))
train_cfg = dict(
    type='EpochBasedTrainLoop', max_epochs=1200, val_interval=20.0)
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')
param_scheduler = [dict(type='PolyLR', power=0.9, eta_min=1e-07, end=1200)]
train_dataloader = dict(
    batch_size=16,
    num_workers=8,
    pin_memory=True,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type='OCRDataset',
        data_root='data/TT_train',
        ann_file='instances_training.json',
        filter_cfg=dict(filter_empty_gt=True, min_size=32),
        pipeline=[
            dict(
                type='LoadImageFromFile',
                color_type='color_ignore_orientation'),
            dict(
                type='LoadOCRAnnotations',
                with_polygon=True,
                with_bbox=True,
                with_label=True),
            dict(type='FixInvalidPolygon', min_poly_points=4),
            dict(
                type='TorchVisionWrapper',
                op='ColorJitter',
                brightness=0.12549019607843137,
                saturation=0.5),
            dict(
                type='ImgAugWrapper',
                args=[['Fliplr', 0.5], {
                    'cls': 'Affine',
                    'rotate': [-10, 10]
                }, ['Resize', [0.5, 3.0]]]),
            dict(type='RandomCrop', min_side_ratio=0.1),
            dict(type='Resize', scale=(640, 640), keep_ratio=True),
            dict(type='Pad', size=(640, 640)),
            dict(
                type='PackTextDetInputs',
                meta_keys=('img_path', 'ori_shape', 'img_shape'))
        ]))
val_dataloader = dict(
    batch_size=1,
    num_workers=1,
    pin_memory=True,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='OCRDataset',
        data_root='data/TT_test',
        ann_file='instances_test.json',
        test_mode=True,
        pipeline=[
            dict(
                type='LoadImageFromFile',
                color_type='color_ignore_orientation'),
            dict(type='Resize', scale=(1333, 736), keep_ratio=True),
            dict(
                type='LoadOCRAnnotations',
                with_polygon=True,
                with_bbox=True,
                with_label=True),
            dict(type='FixInvalidPolygon', min_poly_points=4),
            dict(
                type='PackTextDetInputs',
                meta_keys=('img_path', 'ori_shape', 'img_shape',
                           'scale_factor'))
        ]))
test_dataloader = dict(
    batch_size=1,
    num_workers=1,
    pin_memory=True,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='OCRDataset',
        data_root='data/TT_test',
        ann_file='instances_test.json',
        test_mode=True,
        pipeline=[
            dict(
                type='LoadImageFromFile',
                color_type='color_ignore_orientation'),
            dict(type='Resize', scale=(1333, 736), keep_ratio=True),
            dict(
                type='LoadOCRAnnotations',
                with_polygon=True,
                with_bbox=True,
                with_label=True),
            dict(type='FixInvalidPolygon', min_poly_points=4),
            dict(
                type='PackTextDetInputs',
                meta_keys=('img_path', 'ori_shape', 'img_shape',
                           'scale_factor'))
        ]))
auto_scale_lr = dict(base_batch_size=16)
launcher = 'none'
work_dir = '_test_test/TT'

2023/06/13 16:32:54 - mmengine - INFO - Distributed training is not used, all SyncBatchNorm (SyncBN) layers in the model will be automatically reverted to BatchNormXd layers if they are used.

2023/06/13 16:32:55 - mmengine - WARNING - The prefix is not set in metric class DumpResults.
2023/06/13 16:32:55 - mmengine - INFO - ModulatedDeformConvPack backbone.layer2.0.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer2.1.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer2.2.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer2.3.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer3.0.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer3.1.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer3.2.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer3.3.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer3.4.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer3.5.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer4.0.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer4.1.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - ModulatedDeformConvPack backbone.layer4.2.conv2 is upgraded to version 2.
2023/06/13 16:32:56 - mmengine - INFO - Load checkpoint from _ckpt_new/TT/epoch_1200.pth
2023/06/13 16:34:02 - mmengine - INFO - Epoch(test) [100/300]    eta: 0:02:12  time: 0.6162  data_time: 0.5295  memory: 418  
2023/06/13 16:35:07 - mmengine - INFO - Epoch(test) [200/300]    eta: 0:01:05  time: 0.8017  data_time: 0.7269  memory: 429  
2023/06/13 16:35:57 - mmengine - INFO - Epoch(test) [300/300]    eta: 0:00:00  time: 0.3530  data_time: 0.2867  memory: 429  
2023/06/13 16:35:57 - mmengine - INFO - Evaluating hmean-iou...
2023/06/13 16:35:57 - mmengine - INFO - prediction score threshold: 0.30, recall: 0.7584, precision: 0.6597, hmean: 0.7056

2023/06/13 16:35:57 - mmengine - INFO - prediction score threshold: 0.40, recall: 0.7584, precision: 0.6814, hmean: 0.7178

2023/06/13 16:35:57 - mmengine - INFO - prediction score threshold: 0.50, recall: 0.7584, precision: 0.7060, hmean: 0.7313

2023/06/13 16:35:57 - mmengine - INFO - prediction score threshold: 0.60, recall: 0.7545, precision: 0.7346, hmean: 0.7444

2023/06/13 16:35:57 - mmengine - INFO - prediction score threshold: 0.70, recall: 0.7396, precision: 0.7730, hmean: 0.7559

2023/06/13 16:35:57 - mmengine - INFO - prediction score threshold: 0.80, recall: 0.6740, precision: 0.8286, hmean: 0.7433

2023/06/13 16:35:57 - mmengine - INFO - prediction score threshold: 0.90, recall: 0.3401, precision: 0.9068, hmean: 0.4947

2023/06/13 16:35:57 - mmengine - INFO - Results has been saved to _test_test/TT/epoch_1200.pth_predictions.pkl.
2023/06/13 16:35:57 - mmengine - INFO - Epoch(test) [300/300]    icdar/precision: 0.7730  icdar/recall: 0.7396  icdar/hmean: 0.7559  data_time: 0.5208  time: 0.6004

### Suggest a potential alternative/fix

could you provide a total_converter.py?
or at least provide a mini_totaltext dataset to show us what is the correct format of the curved text dataset?

**Comment by gaotongxiao** [comment]
Totaltext is already supported in Dataset Preparer: https://mmocr.readthedocs.io/en/latest/user_guides/data_prepare/datasetzoo.html

And you should be able to reproduce the result in https://mmocr.readthedocs.io/en/latest/textdet_models.html#total-text with the dataset.

**Comment by tzm-tora** [comment]
Thank you for your quick reply, I have reproduced the result!
